### PR TITLE
Improve ellipsising of comments and add slide open animation

### LINKED ISSUE #3610: Collapsed emoji comment shows tofu
When a comment containing emoji is collapsed with an ellipsis, the last emoji might become tofu.

### Version

- 0.19.3

### Steps to reproduce the bug

1. Open https://www.youtube.com/watch?v=HCjNJDNzw8Y
2. Look at the first comment (currently by KEYERDEY 1 year ago with 37655 votes). Notice the tofu on the second line before the ellipsis
3. Tap on it
4. Look at it again. Notice the second line displays properly now

### Expected behavior

To not see tofu.

For the second line to look like this:

```
🌑🌑🌑🌑🌑🌑🌑🌑🌑🌑
🌑🌑🌑🌑🌑🌑🌑🌑🌑🌑…
```

YouTube web properly displays it like the following (but I am not suggesting changing the number of lines):

```
🌑🌑🌑🌑🌑🌑🌑🌑🌑🌑
🌑🌑🌑🌑🌑🌑🌑🌑🌑🌑
🌑🌑🌓🌕🌕🌕🌕🌕🌗🌑
🌑🌑🌓🌕🌕🌕🌕🌕🌗🌑…
```

### Actual behaviour

I see tofu.

The last moon on the second line is corrupted:

```
🌑🌑🌑🌑🌑🌑🌑🌑🌑🌑
🌑🌑🌑🌑🌑🌑🌑🌑🌑�…
```

This bug might also affect any other Unicode grapheme that (like emoji) is longer than 2 bytes.

### Screenshots/Screen recordings

The comment with emoji is damaged near the ellipsis when collapsed:

Screenshot_20200518-005526

The bug is with the ellipsis because the comment displays fine when expanded:

Screenshot_20200518-005529

### LINKED ISSUE #4009: Animation for expanding video description is broken
### Version
- 0.19.8

### Steps to reproduce the bug
1. Go to a video
2. Tap on video title

### Expected behavior
The animation when touching video title to expand the description is smooth and doesn't feel janky.

### Actual behavior
The animation when tapping the video title seems cut short/laggy. It causes a circle to appear instantly around where you tapped and it then goes away too fast.

### LINKED ISSUE #7334: Add animation when opening more features (Share, etc.) on comment page
### Checklist

- [x] I checked, but didn't find any duplicates (open OR closed) of this issue in the repo. 
- [x] I have read the contribution guidelines given at https://github.com/TeamNewPipe/NewPipe/blob/HEAD/.github/CONTRIBUTING.md.
- [x] This issue contains only one feature request. I will open one issue for every feature I want to request.

#### What feature do you want?

When clicking on the dropdown menu to share the video --> 

#### Why do you want this feature?

Because the UI could become more smooth :)

### PR DESCRIPTION
**Pull request by devlearner**
#### What is it?
- [x] Bugfix (user facing)
- [x] Feature (user facing)
- [ ] Codebase improvement (dev facing)
- [ ] Meta improvement to the project (dev facing)

#### Before/After Screenshots/Screen Record

|         Before          |       After         |
| ------------------- | --------------- |
|  |  |

(The majority of changes go to `NewPipeTextView` (our sub-classed `TextView`) at widget level; while I'm at it, I've also conveniently applied an animation to the Secondary controls panel in `VideoDetailFragment`)

#### Fixes the following issue(s)

- Fixes #3610
- Fixes #4009
- Fixes #7334

_p.s._ I just realize afterwards that there used to be an `AnimationUtil` class which was ported to Kotlin extensions a while ago (by one of our great developers), the 'revival' of `AnimationUtil` was a pure coincidence and with all due respect unintended (but it also does a couple of things extra like keeping a list of running and trying to reuse old animators); probably need to ponder how to better integrate with the existing `animate(...)()` ktx's, but I'll just leave it as is in draft for now

_p.s.(2)_ I saw there're a couple of brilliant PRs in flight which also touch upon `NewPipeTextView`; I'll rebase this PR after they're merged (which I guess would be pretty soon)

#### APK testing

The APK can be found by going to the "Checks" tab below the title. On the left pane, click on "CI", scroll down to "artifacts" and click "app" to download the zip file which contains the debug APK of this PR.

#### Due diligence
- [x] I read the [contribution guidelines](https://github.com/TeamNewPipe/NewPipe/blob/HEAD/.github/CONTRIBUTING.md).

### CODE REVIEW

**Comment by opusforlife2** [comment]
Don't put technical details in a PR title.

**Comment by opusforlife2** [comment]
> ... amidst a cold night

You think you can manage finishing the PR before the gloom overtakes? xD

**Comment by devlearner** [comment]
(oh I thought nobody actually reads the (extended) commit messages... xD ~that was an ode from literature class~)
I was thinking of wrapping this up and calling it a year, but we'll see ... xD

File: app/src/main/java/org/schabi/newpipe/fragments/detail/VideoDetailFragment.java [hunk 1]
```diff
@@ -597,19 +599,36 @@ public boolean onLongClick(final View v) {
...
+                        VideoPlayerUi.DEFAULT_CONTROLS_DURATION, 0);
+                AnimationUtil.collapse(binding.detailSecondaryControlPanel, 500,
+                        (animatedFraction, isCollapsing) -> updateTabLayoutVisibility());
+            }
         } else {
-            binding.detailVideoTitleView.setMaxLines(1);
-            animateRotation(binding.detailToggleSecondaryControlsView,
-                    VideoPlayerUi.DEFAULT_CONTROLS_DURATION, 0);
-            binding.detailSecondaryControlPanel.setVisibility(View.GONE);
+            // legacy approach
```
Comment (Stypox) [inline app/src/main/java/org/schabi/newpipe/fragments/detail/VideoDetailFragment.java:617]: I think you can safely assume that `binding.detailVideoTitleView` will always be instance of `NewPipeTextView`. Also, I wonder why the binding element is not already cast to that type. Maybe check if the text view has type `NewPipeTextView` both in the normal and the large-land layouts

**Review by Stypox — CHANGES_REQUESTED** [review]
Thanks, the video detail fragment animation looks really nice! And also the ellipsizing process. This is a fast review just to understand better what your code does, I get it that it's still in draft state.
Would it be possible to split the two features into two separate PRs? That would make merging them easier and faster.
Did you write all of the code in `EllipsizeParams` and `AnimationUtil`, or did you take it from some system class? It seems really complicated for what it is meant to do :sweat_smile:.
Could you expand on how you fixed the comment problems? If I understand correctly you now paint the three dots on top of the trimmed text, to make sure the text is not trimmed at half a character and at the same time ensure the text does not overflow?

**Comment by Stypox** [comment]
Closing as there has been no recent activity, and we're going to rewrite this part with Jetpack Compose which is going to make things much simpler. Thank you anyway!

**Pull request closed without merging**